ui/src/views/WorkerManager/WMEditWorkerPool/index.jsx

```diff
@@ -0,0 +1,203 @@
+import { bool } from 'prop-types';
+import TextField from '@material-ui/core/TextField/TextField';
+import FormGroup from '@material-ui/core/FormGroup/FormGroup';
+import FormControlLabel from '@material-ui/core/FormControlLabel/FormControlLabel';
+import FormLabel from '@material-ui/core/FormLabel/FormLabel';
+import MenuItem from '@material-ui/core/MenuItem/MenuItem';
+import CodeEditor from '@mozilla-frontend-infra/components/CodeEditor';
+import ListItem from '@material-ui/core/ListItem/ListItem';
+import Switch from '@material-ui/core/Switch/Switch';
+import { withStyles } from '@material-ui/core';
+import CheckIcon from 'mdi-react/CheckIcon';
+import List from '../../Documentation/components/List';
+import isWorkerTypeNameValid from '../../../utils/isWorkerTypeNameValid';
+import Dashboard from '../../../components/Dashboard';
+import Button from '../../../components/Button';
+
+const gcpConfig = {
+  minCapacity: 0,
+  maxCapacity: 0,
+  capacityPerInstance: 1,
+  machineType: 'n1-highcpu-8',
+  regions: ['us-west2'],
+  userData: {},
+  scheduling: {},
+  networkInterfaces: [{}],
+  disks: [{}],
+};
+
+@hot(module)
+@withStyles(theme => ({
+  successIcon: {
+    ...theme.mixins.successIcon,
+  },
+  createIconSpan: {
+    ...theme.mixins.fab,
+    ...theme.mixins.actionButton,
+    right: theme.spacing.unit * 11,
+  },
+}))
+export default class WMWorkerPoolEditor extends Component {
+  static defaultProps = {
+    isNewWorkerPool: true,
+  };
+
+  static propTypes = {
+    isNewWorkerPool: bool,
+  };
+
+  state = {
+    workerPool: {
+      name: '',
+      description: '',
+      owner: '',
+      wantsEmail: false,
+      providerType: 'gcp',
+      providerId: '',
+      config: gcpConfig,
+    },
+    invalidProviderConfig: false,
+  };
+
+  handleInputChange = ({ target: { name, value } }) => {
+    this.setState({ workerPool: { ...this.state.workerPool, [name]: value } });
+  };
+
+  handleSwitchChange = event => {
+    const {
+      target: { value },
+    } = event;
+
+    this.setState({
+      workerPool: {
+        ...this.state.workerPool,
+        [value]: !this.state.workerPool[value],
+      },
+    });
+  };
+
+  handleEditorChange = value => {
+    const { workerPool } = this.state;
+
+    try {
+      workerPool.config = JSON.parse(value);
+
+      this.setState({
+        workerPool,
+        invalidProviderConfig: false,
+      });
+    } catch (err) {
+      workerPool.config = value;
+
+      this.setState({
+        workerPool,
+        invalidProviderConfig: true,
+      });
+    }
+  };
+
+  render() {
+    const { isNewWorkerPool, classes } = this.props;
+    const { workerPool, invalidProviderConfig } = this.state;
+
+    return (
+      <Dashboard
+        title={
+          isNewWorkerPool
+            ? 'Worker Manager: Create Worker Type'
+            : 'Worker Manager: Edit Worker Type'
+        }>
+        <TextField
+          label="Enter Worker Type Name..."
```

Comment:
(wow, I missed a spot there - should have been `worker pool`)

That label is also a place holder, and for me personally it's always a source of confusion if there's no clear indication about it being a placeholder. It always takes me several seconds to realize where I have to enter information on forms like that. Usually it doesn't even register with me immediately that I am looking at a form. Maybe it's just me, of course, but in any case, my intention was to make as clear as possible that this is a placeholder, and make it inviting to click on it and enter things

Reply:
Agreed, it could be a source of confusion. To make it clearer, we could make use of a helper text. I think the label should simply be saying the "what" otherwise once a user fills in the input or if the user is viewing an existent form, the UI might be confusing to the end user. It will look like

![Screen Shot 2019-05-31 at 1 54 19 PM](https://user-images.githubusercontent.com/3766511/58725038-27642980-83ac-11e9-970f-dc0963dec903.png)

Making use of a helper text will hopefully remove the confusion.

![Screen Shot 2019-05-31 at 1 54 04 PM](https://user-images.githubusercontent.com/3766511/58725184-86c23980-83ac-11e9-80e3-aedfd704fa53.png)

Does that help?

Reply:
helper text is a tooltip? I already use that. It can't be replacement for a placeholder because tooltips don't work on mobile as there's no "hover" action on mobile

What I fail to understand is why a label is also a placeholder lol but I guess that's a rant about the specific library we use. The first screenshot looks better to me tho than opening a form and seeing labels with no fields only to realize in a couple of seconds that I should enter text where the labels are

Reply:
Helper text is not a tooltip. It's an actual prop for the `TextField` component. It will show as small text below the text value. When the text input is empty, it will look like:

![Screen Shot 2019-05-31 at 3 52 48 PM](https://user-images.githubusercontent.com/3766511/58731215-2e933380-83bc-11e9-91a1-9b5b8c8d54e5.png)

When the input is filled, it will look like:

![Screen Shot 2019-05-31 at 1 54 04 PM](https://user-images.githubusercontent.com/3766511/58725184-86c23980-83ac-11e9-80e3-aedfd704fa53.png)

```jsx
<TextField
  helperText="Please enter a Worker Pool name"
  label="Worker Pool"
```

Reply:
oh, ok, that makes more sense. However, I don't like it because now there's too much text this way. I'd rather not use this.

The thing about a placeholder is that it's in the component when you haven't clicked into the component yet, and it's not there anymore once you start typing. This component just completely lacks such element! Cannot say I understand these design choices in this library, but between good looks and functionality/easy-to-use I'd choose the latter. I still like this the best:

![Screen Shot 2019-05-31 at 1 54 19 PM](https://user-images.githubusercontent.com/3766511/58725038-27642980-83ac-11e9-970f-dc0963dec903.png)

Because it has less text than the third option, and guides user better than the second one

Reply:
It would be great to have this part of the UI "blend in" with the other parts.  I agree that material design has some weird and difficult-to-use aspects, but at least the UI is consistent about it.  For example, in hooks:

![image](https://user-images.githubusercontent.com/28673/58735900-0ad6ea00-83ca-11e9-8d9a-aefc43ace35d.png)

or when creating a new hook:

![image](https://user-images.githubusercontent.com/28673/58735930-1aeec980-83ca-11e9-8335-c704d69a5e10.png)

I don't think there's any reason to treat "Hook ID" differently from "Worker Pool ID".

If we want to address the placeholder usability issue across the whole UI, that might make sense -- but let's do it in another PR rather than fixing it partially in this PR.

Reply:
I would at least have the `...` in there to indicate that this is a clickable field. I am remembering now that you have to have a `Please` in English so that it sounds polite, and with `Please enter` that might be too much text. Maybe `...` alone would do the job.

As far as consistency goes, I volunteer to make this change on other forms, because this is something I'd like to change about our forms. I agree that can be a separate PR tho

Reply:
Feel free to omit the "Please". It's okay for a helper text to look like an order just like our tooltips.

Reply:
> As far as consistency goes, I volunteer to make this change on other forms, because this is something I'd like to change about our forms. I agree that can be a separate PR tho

Feel free to create an issue on the repo with the suggestion and we can try to find a consensus. One thing we could do is have the first text field be focused on page load when the user navigates to it. This will make sure there is no confusion that it is a placeholder.

Reply:
> One thing we could do is have the first text field be focused on page load when the user navigates to it.

This is an excellent suggestion, actually. Ok, for now I'll maintain consistency with the other forms, and I'll create an issue about this - let's discuss some more outside of this PR.